### PR TITLE
Update python ecosystem dependencies

### PR DESCRIPTION
**Pull request by chris48s**
PR #51 was opened and then abandoned. It has not been touched in some months.

While well intentioned, one of the problems with that PR is it tried to a lot of things at once. It was:

- Updating python and related packages
- Updating javascript packages
- Implementing feature https://github.com/wagtail/cookiecutter-wagtail-package/issues/14
- Migrating linters from black/isort/flake8 to ruff

Those weren't necessarily bad or wrong things to do, but doing them all at once made it hard to review and also hard to pick up the PR and update it or get it finished.

In this PR, I have intentionally reduced the scope. All I am doing here is updating python and python ecosystem-related packages. I am intentionally not touching anything node/javascript related in this PR and I'm sticking with black/isort/flake8. Hopefully this will make it easier to get this reviewed and merged. We can look at the other issues in further more tightly scoped PRs

Closes https://github.com/wagtail/cookiecutter-wagtail-package/issues/50

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -29,7 +29,7 @@ jobs:
           python-version: '3.11'
 
       - uses: actions/download-artifact@v3
-        # Restore the rendered areifact to a directory outside the repo checkout.
+        # Restore the rendered artifact to a directory outside the repo checkout.
```
Comment (chris48s) [inline .github/workflows/test.yml:32]: Tiny bit of scope creep here just to fix a typo.
I am going with the en-US spelling "artifact" here (in preference to the en-GB spelling "artefact") as that is the spelling GitHub uses.

**Comment by chris48s** [comment]
The build on this PR is failing due to an ESLint error.

This is unrelated to changes in this PR. I've submitted https://github.com/wagtail/cookiecutter-wagtail-package/pull/53 just to prove that point. That PR has no functional changes from the tip of `main` and reproduces the same issue.

I guess I'm going to need to scope-creep this to include the minimum amount of js/pre-commit faffing needed to get the build passing again. Not today though. Not now.

My gut instinct is that there is an implicit eslint-related dependency which is unpinned in https://github.com/wagtail/cookiecutter-wagtail-package/blob/5625bcf0f3df2a37c5bc65c88dfe185cfc51edda/%7B%7B%20cookiecutter.__project_name_kebab%20%7D%7D/.pre-commit-config.yaml#L51-L60 but I will put a pause on grumbling about pre-commit making us re-invent lockfiles until I've worked out if that is actually the problem.

File: {{ cookiecutter.__project_name_kebab }}/.github/workflows/test.yml [hunk 2]
```diff
@@ -67,20 +69,19 @@ jobs:
     needs: lint
     strategy:
       matrix:
-        python: ['3.7', '3.8', '3.9']
-        django: ['3.2']
-        wagtail: ['2.15', '4.0', '4.1']
+        python: ['3.8', '3.9', '3.10', '3.11']
+        django: ['3.2', '4.0', '4.1', '4.2']
+        wagtail: ['4.1', '4.2', '5.0', '5.1']
```
Comment (zerolab) [inline {{ cookiecutter.__project_name_kebab }}/.github/workflows/test.yml:74]: it is a pity we have the tox config, and then this. We should follow up with setup for https://github.com/ymyzk/tox-gh-actions/ so that we only run tox in CI

**Review by zerolab — APPROVED** [review]

File: {{ cookiecutter.__project_name_kebab }}/.github/workflows/test.yml (hunk 2 above)
Comment (chris48s) [reply {{ cookiecutter.__project_name_kebab }}/.github/workflows/test.yml:74]: I've made an issue for this https://github.com/wagtail/cookiecutter-wagtail-package/issues/55
